src/fp.rs

```diff
@@ -0,0 +1,367 @@
+        lo = lo64(result);
+        result = lo + carry;
+        lo = lo64(result);
+        cc = hi64(result);
+        result = hi + cc;
+        hi = lo64(result);
+        result = zz[2] + lo;
+        zz[2] = lo64(result);
+        cc = hi64(result);
+        result = zz[3] + hi + cc;
+        zz[3] = lo64(result);
+
+        zz[2] | (zz[3] << 64)
+    }
+
+    /// Modular exponentiation, i.e., `x^exp (mod p)` where `p` is the modulus. Note that the
+    /// runtime of this algorithm is linear in the bit length of `exp`.
+    pub fn pow(&self, x: u128, exp: u128) -> u128 {
+        let mut t = self.elem(1);
+        for i in (0..128).rev() {
+            t = self.mul(t, t);
+            if (exp >> i) & 1 != 0 {
+                t = self.mul(t, x);
+            }
+        }
+        t
+    }
+
+    /// Modular inversion, i.e., x^-1 (mod p) where `p` is the modulu. Note that the runtime of
+    /// this algorithm is linear in the bit length of `p`.
+    pub fn inv(&self, x: u128) -> u128 {
+        self.pow(x, self.p - 2)
+    }
+
+    /// Negation, i.e., `-x (mod p)` where `p` is the modulus.
+    pub fn neg(&self, x: u128) -> u128 {
+        self.sub(0, x)
+    }
+
+    /// Maps an integer to its internal representation. Field elements are mapped to the Montgomery
+    /// domain in order to carry out field arithmetic.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use prio::fp::FieldParameters;
+    /// let fp = FieldParameters { p: 4293918721, p2: 8587837442, mu: 17302828673139736575, r2: 1676699750 };
+    /// let integer = 1; // Standard integer representation
+    /// let elem = fp.elem(integer); // Internal representation in the Montgomery domain
+    /// assert_eq!(elem, 2564090464);
+    /// ```
+    pub fn elem(&self, x: u128) -> u128 {
+        modp(self.mul(x, self.r2), self.p)
+    }
+
+    /// Returns a random field element mapped.
+    pub fn rand_elem<R: Rng + ?Sized>(&self, rng: &mut R) -> u128 {
+        let uniform = rand::distributions::Uniform::from(0..self.p);
+        self.elem(uniform.sample(rng))
+    }
+
+    /// Maps a field element to its representation as an integer.
+    /// ```
+    /// use prio::fp::FieldParameters;
+    /// let fp = FieldParameters { p: 4293918721, p2: 8587837442, mu: 17302828673139736575, r2: 1676699750 };
+    /// let elem = 2564090464; // Internal representation in the Montgomery domain
+    /// let integer = fp.from_elem(elem); // Standard integer representation
+    /// assert_eq!(integer, 1);
+    /// ```
+    pub fn from_elem(&self, x: u128) -> u128 {
+        modp(self.mul(x, 1), self.p)
+    }
+
+    /// Returns the number of bytes required to encode field elements.
+    pub fn size(&self) -> usize {
+        (16 - (self.p.leading_zeros() / 8)) as usize
+    }
+
+    #[cfg(test)]
+    pub fn check(&self) -> Result<(), &'static str> {
+        use modinverse::modinverse;
+        use num_bigint::{BigInt, ToBigInt};
+
+        let err_modulus_too_large = "p > 2^126";
```

Comment:
nit: this is fine for now since the only caller of this method is only going to print any error they encounter, but if we end up with more errors from `fp`, it'd be best to define an `fp::Error` enum with variants for specific error cases (like [finite_field does](https://github.com/abetterinternet/libprio-rs/blob/e0247d036716f3bdbe8702147568a7e107dc5d28/src/finite_field.rs#L8)).

Reply:
Ack. Yeah, this seems like it will be handy. I'll keep this in mind for the future.